[#40228] added proposal for openapi spec

Kharonus opened the pull request:
- fixed some warnings and spec errors in work_package_watchers.yml and
  work_package.yml on the way
- removed unused component schemas

https://community.openproject.org/wp/40228

Kharonus committed:
[#40228] added proposal for openapi spec

- fixed some warnings and spec errors in work_package_watchers.yml and
  work_package.yml on the way
- removed unused component schemas

Kharonus committed:
[#40228] reworked creation request

- use common schema for origin data
- collection as request body for link creation

Kharonus committed:
[#40228] added work packages filter docs

ulferts commented on docs/api/apiv3/components/schemas/file_link_model.yml:
```diff
@@ -0,0 +1,138 @@
+      - FileLink
+  storageType:
+    type: string
+    readOnly: true
```
Why is this property part of the `FileLinkl` resource? Isn't it rather a property of the storage?

Kharonus replied:
Are you worried about duplication of information and thus a potential error source? For the db schema I'd approve (type is only stored on the storage, not on the file link), here I was thinking about the possible filtering of the file links. In the beginning we listed the `storage_type` as a necessary filter for `file_links` endpoint.

Kharonus replied:
Moved to storage resource

ulferts commented on docs/api/apiv3/components/schemas/file_link_model.yml:
```diff
@@ -0,0 +1,138 @@
+  _type:
+    type: string
+    enum:
+      - FileLink
```
This is also a `readOnly` property. Additionally, I'd have thought it should be possible to use `const` but since we are still on OpenAPI 3.0, I guess `enum` is just fine.

Kharonus replied:
Do you have information about `const` being supported by openapi 3.1? cannot find it here: https://spec.openapis.org/oas/v3.1.0

added `readOnly`

Kharonus replied:
- keeping enum instead of const, as we need 3.0.x compatibility
- reworked `readOnly` and `writeOnly` on the new models

ulferts commented on docs/api/apiv3/components/schemas/file_link_model.yml:
```diff
@@ -0,0 +1,138 @@
+              Can be one of the following **Resources**: 
+              
+              - WorkPackage
+            readOnly: true
```
Why is this `readOnly`? Isn't this link how you set the connection between the container and the `FileLink`?

Kharonus replied:
This link should be the one pointing to the container itself, in the very first iteration it is a getter on a work package.

Kharonus replied:
reworked readOnly and writeOnly on the new models

ulferts commented on docs/api/apiv3/components/schemas/file_link_model.yml:
```diff
@@ -0,0 +1,138 @@
+              The storage url the file link references to.
+              
+              **Resource**: N/A
+            readOnly: true
```
This should be writable since the client needs to set the storage. But it should be restricted to the set of storages configured for integration in the OpenProject instance (if that is part of the MVP).

If the configuration of the available storages is part of the MVP, this would suggest to have `Storage` resources which would then also contain the `storageType` property.

Kharonus replied:
I thought about it, and maybe I do not fully understand yet the `_link` property. In my mind, it was a readonly collection of consumable links, for file link creation I defined out of this thought a property `storage_url`, which lies then on the same layer then the `storage_type`

Kharonus replied:
reworked readOnly and writeOnly on the new models

ulferts commented on docs/api/apiv3/components/schemas/file_link_model.yml:
```diff
@@ -0,0 +1,138 @@
+              The uri to open the origin file on the origin itself
+              
+              **Resource**: N/A
+            readOnly: true
```
I notice there is no `update` link which is probably intended. That brought me to think about the intention of a `readOnly` property. From my understanding, all of the properties the client has to provide, and that is `originData`, `storage` and `container`, cannot be `readOnly` even though they can never be updated. But they can only not be updated because there is no end point to do so. I wonder though, how one would have to specify a property which upon creation is writable but which after that cannot be altered any more while other properties of the same resource can.

Kharonus replied:
reworked readOnly and writeOnly on the new models

ulferts commented on docs/api/apiv3/paths/work_package_file_link.yml:
```diff
@@ -0,0 +1,119 @@
+          example:
+            _type: Error
+            errorIdentifier: urn:openproject-org:api:v3:errors:MissingPermission
+            message: You are not allowed to delete file links of this work package.
```
We currently only provide a default message without any context specific information. This could be improved upon but doing so will require additional work. 

Not saying that I am against it, but I wanted to warn you.

Kharonus replied:
changed to default message

ulferts commented on docs/api/apiv3/paths/work_package_file_link.yml:
```diff
@@ -0,0 +1,119 @@
+      description: |-
+        Returned if the work package does not exist or the client does not have sufficient permissions to see it.
+
+        **Required permission:** view work package
```
This could also happen if no `FileLink` with the provided id exists (for this work package).

Kharonus replied:
amended required permissions

ulferts commented on docs/api/apiv3/paths/work_package_file_link.yml:
```diff
@@ -0,0 +1,119 @@
+          example:
+            _type: Error
+            errorIdentifier: urn:openproject-org:api:v3:errors:MissingPermission
+            message: You are not allowed to delete file links of this work package.
```
See above

Kharonus replied:
changed to default message

ulferts commented on docs/api/apiv3/paths/work_package_file_link.yml:
```diff
@@ -0,0 +1,119 @@
+      description: |-
+        Returned if the work package does not exist or the client does not have sufficient permissions to see it.
+
+        **Required permission:** view work package
```
See above

Kharonus replied:
amended required permissions

ulferts commented on docs/api/apiv3/paths/work_package_file_link_download.yml:
```diff
@@ -0,0 +1,62 @@
+    - File links
+  description: |-
+    Creates a download uri of the origin file linked by the given file link. This uri depends on the storage type and
+    always is located on the origin storage itself.
```
Can you please elaborate on why that is necessary?

And is that link actually created or rather fetched from the server. I would tend to the later and hence would prefer a GET over the POST. Additionally, we could return a redirect which would make it easier for the clients to follow.

Kharonus replied:
I'd say, the URI points always to the storage linked inside the file link, as the whole concept is built around the idea, of just saving a reference to the original file.

As for the idea, of instead of making a response containing the URI where to find that file, or directly returning a redirect, I'm just not sure, what we want as a behavior of our API.

Kharonus replied:
reworked static resource endpoints to return redirects

ulferts commented on docs/api/apiv3/paths/work_package_file_link_download.yml:
```diff
@@ -0,0 +1,62 @@
+          example:
+            _type: Error
+            errorIdentifier: urn:openproject-org:api:v3:errors:MissingPermission
+            message: You are not allowed to delete file links of this work package.
```
This message is not suitable here.

Kharonus replied:
made more suitable

ulferts commented on docs/api/apiv3/paths/work_package_file_link_open.yml:
```diff
@@ -0,0 +1,62 @@
+    - File links
+  description: |-
+    Creates a uri to open the origin file linked by the given file link. This uri depends on the storage type and
+    always is located on the origin storage itself.
```
See above

Kharonus replied:
explained ;)

ulferts commented on docs/api/apiv3/paths/work_package_file_links.yml:
```diff
@@ -0,0 +1,212 @@
+          required:
+            - _type
+            - count
+            - total
```
we could declare the `count` and `total` properties to be `readOnly` as I don't see the value in the client sending this information, especially both.

Kharonus replied:
ah, copy-paste error of mine. As this is the request body, we could completely leave those properties out. If we say, they are part of the collection model, and they are marked as `readOnly`, which component would set them instead?

Kharonus replied:
changed request body properties

ulferts commented on docs/api/apiv3/paths/work_package_file_links.yml:
```diff
@@ -0,0 +1,212 @@
+          example:
+            _type: Error
+            errorIdentifier: urn:openproject-org:api:v3:errors:MissingPermission
+            message: You are not allowed to create file links of this work package.
```
The message is incorrect.

Kharonus replied:
made it correct

ulferts commented on docs/api/apiv3/paths/work_packages.yml:
```diff
@@ -2,57 +2,112 @@
+        - updated_at
+        - version
+        - watcher
+        - work_package
```
Thanks a ton @Kharonus

ulferts commented on docs/api/apiv3/tags/file_links.yml:
```diff
@@ -0,0 +1,4 @@
+---
+description: |-
+  TBD
```
We should add something here at least when the feature has been developed.

Kharonus replied:
Acknowledged. But in this very early state I have the feeling, that every word I'd write in here, I'd need to replace in 2 weeks from here.

ulferts reviewed (changes requested):
@Kharonus general structure is good so no major roadblocks.

Kharonus committed:
[#40228] added storage resource

- changed /download and /open to redirect responses
- fixed readOnly properties
- added storage path and model
- added permissions and descriptions on error responses

ulferts commented on docs/api/apiv3/components/schemas/file_link_model.yml:
```diff
@@ -30,6 +27,17 @@ properties:
+        $ref: '#/components/schemas/StorageModel'
+      container:
+        oneOf:
+          - $ref: '#/components/schemas/Work_PackageModel'
```
Both of those are `readOnly: true`

Kharonus replied:
put `readOnly` on the `_embedded` itself, as it is not feasible to put it next to a `$ref`.

ulferts commented on docs/api/apiv3/components/schemas/file_link_model.yml:
```diff
@@ -45,18 +53,26 @@ properties:
+            writeOnly: true
       container:
         allOf:
           - $ref: './link.yml'
```
`container` should be allowed to be written at least on the creation.

Kharonus replied:
Why is it needed? The container is defined on the route of the `POST`. So, it is not needed in the request body model.

Kharonus replied:
resolved

ulferts commented on docs/api/apiv3/components/schemas/file_link_model.yml:
```diff
@@ -114,13 +137,163 @@ example:
+    storage:
+      id: 1337
+      _type: Storage
+      href: https://nextcloud.deathstar.rocks/
```
Why not as part of the `_links`-section?

Kharonus replied:
moved to `_links` section

ulferts commented on docs/api/apiv3/components/schemas/file_link_model.yml:
```diff
@@ -114,13 +137,163 @@ example:
+          href: /api/v3/storages/1337
+          title: storage
+    container:
+      _type: WorkPackage
```
Any chance we can `$ref` that?

Kharonus replied:
`$ref` references objects, our type is a string. not only in this model, but overall. Any good reason to work on changing that?

Kharonus replied:
referenced the single examples

Kharonus replied:
example references doesn't get resolved, so we use now shortened examples in the `_embedded` section

ulferts commented on docs/api/apiv3/components/schemas/storage_model.yml:
```diff
@@ -0,0 +1,59 @@
+  storageType:
+    type: string
+    enum:
+      - nextcloud
```
We could also introduce a uri for this to highlight the fact that this is actually a constant. Then, the storageType would be moved to the `_links` section and could also receive a `title` attribute.

Kharonus replied:
Is there any benefit from moving constants to uris? why is a resource behind a uri more likely a constant, than an enumeration string?

Kharonus replied:
moved to _links as an urn.

ulferts commented on docs/api/apiv3/components/schemas/storage_model.yml:
```diff
@@ -0,0 +1,59 @@
+          - description: |-
+              This storage resource
+
+              **Resource**: Storage
```
`readOnly` properties are not declared.

Kharonus replied:
well, this model is currently used only in a `GET` endpoint, there is no usage in a `POST`, `PATCH` or `PUT`. Hence it is `readOnly` by default. The additional property should IMHO only be used, when a model is used for reading and writing, to differentiate the properties between both scopes.

Kharonus replied:
not needed

ulferts commented on docs/api/apiv3/components/schemas/storage_model.yml:
```diff
@@ -0,0 +1,59 @@
+  _links:
+    self:
+      href: /api/v3/storages/1337
+      title: storage
```
A client defined title/name might be helpful.

Kharonus replied:
With our projects here in self link title is the `Name` of the project. I'd go the same approach with the storage resource, as the user/client defines that name.

Kharonus replied:
But, alas, I noticed, I forgot the storage name in the model. Added.

ulferts commented on docs/api/apiv3/paths/work_package_file_links.yml:
```diff
@@ -62,6 +54,23 @@ post:
+                _links:
+                  storageUrl:
+                    href: https://nextcloud.deathstar.rocks/
+                    title: Link to storage
```
The `title` key is not necessary.

Kharonus replied:
removed

Kharonus committed:
[#40228] fixed PR issues

Kharonus committed:
[#40228] moved storage type property to urn

Kharonus committed:
[#40228] added creator to file link model

Kharonus committed:
[#40228] removed oas test file

ulferts reviewed (approved):
👍

ulferts commented:
Merging even though the unit/features are failing. The reason for this is unrelated to the PR and there are no code changes within this PR. Specification building is working.
